src/net/rpc.rs

```diff
@@ -543,6 +548,131 @@ impl ConversationHttp {
         }
     }
 
+    fn handle_getattachmentsinv<W: Write>(
+        http: &mut StacksHttp,
+        fd: &mut W,
+        req: &HttpRequestType,
+        sortdb: &SortitionDB,
+        atlasdb: &AtlasDB,
+        chainstate: &mut StacksChainState,
+        tip_consensus_hash: &ConsensusHash,
+        tip_block_hash: &BlockHeaderHash,
+        pages_indexes: &HashSet<u32>,
+        options: &ConnectionOptions,
+    ) -> Result<(), net_error> {
+        let response_metadata = HttpResponseMetadata::from(req);
+        let mut pages_indexes = pages_indexes.iter().map(|i| *i).collect::<Vec<u32>>();
+        pages_indexes.sort();
+        let tip = StacksBlockHeader::make_index_block_hash(&tip_consensus_hash, &tip_block_hash);
+
+        let (oldest_page_index, newest_page_index, pages_indexes) = if pages_indexes.len() > 0 {
+            (
+                *pages_indexes.first().unwrap(),
+                *pages_indexes.last().unwrap(),
+                pages_indexes.clone(),
+            )
+        } else {
+            // Pages indexes not provided. By default, we'll return the inventory of the latest page
+            // present on the canonical chain.
+            let res =
+                chainstate.maybe_read_only_clarity_tx(&sortdb.index_conn(), &tip, |clarity_tx| {
+                    let cost_tracker =
+                        LimitedCostTracker::new(options.read_only_call_limit.clone());
+                    OnchainInventoryLookup::get_attachments_inv_info(cost_tracker, clarity_tx)
+                });
+            match res {
+                Ok(inv_info) => {
+                    let current_page = inv_info.pages_count - 1;
+                    (current_page, current_page, vec![current_page])
+                }
+                Err(_) => {
+                    let msg = format!("Unable to read contract BNS");
+                    warn!("{}", msg);
+                    let response = HttpResponseType::ServerError(response_metadata, msg.clone());
+                    response.send(http, fd)?;
+                    return Err(net_error::ChainstateError(msg));
+                }
+            }
+        };
+
+        // We need to rebuild an ancestry tree, but we're still missing some informations at this point
+        let (min_block_height, max_block_height) = match atlasdb
+            .get_minmax_heights_window_for_page_index(oldest_page_index, newest_page_index)
+        {
+            Ok(window) => window,
+            Err(e) => {
+                let msg = format!("Unable to read Atlas DB");
+                warn!("{}", msg);
+                let response = HttpResponseType::ServerError(response_metadata, msg.clone());
+                response.send(http, fd)?;
+                return Err(net_error::DBError(e));
+            }
+        };
```

Comment:
Can this code enforce a maximum difference between `min_block_height` and `max_block_height`?  Otherwise, an attacker could DDoS a node by asking for the _entire_ blockchain range of inventory statuses.

Reply:
Actually, one thing I'd like you to do here (but not in this PR -- please open an issue) is to stream back the inventory response.  The streaming context can do a bound number of MARF reads per invocation (e.g. load one block ancestor) before yielding.

Reply:
For example, one pass of the stream context could load up a single ancestor's `StacksBlockId`, load its bit-vector of attachments for it, and then write just that information to the network response handle.  Successive calls to the stream context could just load up earlier and earlier ancestor `StacksBlockId`s and their bit-vectors.  Then, we don't have to worry about this RPC blocking progress on other sockets, since each socket wake-up will only do _O(1)_ I/O before yielding.

Reply:
Sure! will address in a separate issue https://github.com/blockstack/stacks-blockchain/issues/2077

Reply:
Moving the discussion of your first comment here: https://github.com/blockstack/stacks-blockchain/pull/2013#discussion_r525393381